### DISCUSSION TITLE
Encryption query

### DISCUSSION THREAD

**Discussion by AmyJeanes** [teslamate-org/teslamate#2754] [General]
Hey @adriankumpf just wanted to check to see if the new encryption feature would affect me reading any of the following tables directly i.e. from TeslaMateAgile 🙂:
- `charges`
- `charging_processes`
- `geofences`

My understanding is this is only for the Tesla API tokens but wanted to clarify just in case

**Comment by LelandSindt** [comment]
@MattJeanes the `ENCRYPTION_KEY` changes only effect columns  in the `tokens` table.

**Comment by chillr21** [comment]
Can the encrypted tokens from the tokens table be decrypted using our ENCRYPTION_KEY via a bash script?

I'm currently using them to control my powerwall.

**Reply by LelandSindt** [reply]
Off the top of my head I don't know specific commands line tools you could use... 

This... https://github.com/tobiasehlert/teslamateapi/commit/3f9e8c471bf2477ad7827cc208b9f492f4b0bc38 is how we handled the decryption in TeslaMateAPI...

**Reply by LelandSindt** [reply]
The more I think about it... I would encourage you to work with https://github.com/tobiasehlert/teslamateapi/ to add support for power walls...

**Reply by LelandSindt** [reply]
@tobiasehlert

**Reply by tobiasehlert** [reply]
Sure.. what would you like to have in terms of powerwall commands?
I don't have one and it's probably going to take a couple of years until those get sold in Sweden:-P

**Reply by chillr21** [reply]
That would be great!

In my case the only command I use is the `backup_reserve_percent` - as part of my home automation to control whether i'm charging or discharging the powerwall - i.e. at 2300 when power cost is low use my locations solar forecast for tomorrow morning to calculate if and by how much I would have a shortfall of battery charge & solar generation to see me through my grid electricity suppliers morning peak kWh charge period (0700 - 1100), then command the backup_reserve_percent if required.

The specific curl post I use to command it:
```
curl --max-time 5.0 -s     --header "Authorization: Bearer $APITOKEN"      --header "Content-Type: application/json" \
  --request POST \
  --data   '{"backup_reserve_percent":'$BATTERYPERCENT'}' \
  'https://owner-api.teslamotors.com/api/1/energy_sites/xxxxxx/backup'
```
$APITOKEN having been read from the Teslamate Postgres DB

**Reply by LelandSindt** [reply]
Are the `Powerwalls` and `Energy Sites` endpoints listed here what you would be looking for/do they look comprehensive? 

https://www.teslaapi.io/powerwalls/state-and-settings

**Reply by chillr21** [reply]
Yes I use the Energy Sites endpoint and that list seems comprehensive to me.

In the mean time I've made copies my Teslamate / Postgre dockers and updated the Teslamate copy to the v1.27.0 to test decrytping the access token. Using the golang code referenced above I'm getting a decryption error "panic: cipher: message authentication failed" - not sure what I'm doing wrong. 
I'm adding the token from the postgres DB & encryptionKey directly in code as:
```
data := (`\x010a4145.....`)
encryptionKey := (`testkey.....`)
```
Does the start of the encrypted token look correct? \x010a4145

**Reply by LelandSindt** [reply]
Looks like you are getting the token from `psql` in hex format...  

Here is an example that can handle the hex formatted/encrypted token...  https://gist.github.com/LelandSindt/7f3ebd5e8615d62bb74d321728f5c794

**Reply by chillr21** [reply]
That works perfectly, thanks!
